src/layers/convolutional.ts

```diff
@@ -365,18 +371,47 @@ export abstract class Conv extends Layer {
             `convolution, but received ${JSON.stringify(this.dilationRate)}`);
       }
     }
+  }
 
-    this.activation = getActivation(config.activation);
-    this.useBias = config.useBias == null ? true : config.useBias;
+  protected static verifyConfig(config: BaseConvLayerConfig) {
+    // Check config.kernelSize type and shape.
+    generic_utils.assert(
+        'kernelSize' in config, `required key 'kernelSize' not in config`);
+    if (typeof config.kernelSize !== 'number' &&
+        !generic_utils.checkArrayTypeAndLength(
+            config.kernelSize, 'number', 1, 2))
+      throw new ValueError(
+          `BaseConv expects config.kernelSize to be number or number[] with ` +
+          `length 1 or 2, but received ${JSON.stringify(config.kernelSize)}.`);
+  }
+}
+
+/**
+ * Abstract nD convolution layer.  Ancestor of convolution layers which reduce
+ * across channels, i.e., Conv1D and Conv2D, but not DepthwiseConv2D.
+ */
+export abstract class Conv extends BaseConv {
+  protected readonly filters: number;
+
+  protected kernel: LayerVariable = null;
+
+  // Bias-related properties are stored in the superclass `BaseConv` because all
+  // convolution subclasses use the same configuration parameters to control
+  // bias. Kernel-related properties are defined here rather than in the
+  // superclass because some convolution subclasses use different names and
+  // configuration parameters for their internal kernel state.
+  protected readonly kernelInitializer?: Initializer;
+  protected readonly kernelConstraint?: Constraint;
+  protected readonly kernelRegularizer?: Regularizer;
+
+  constructor(rank: number, config: ConvLayerConfig) {
+    super(rank, config as BaseConvLayerConfig);
+    Conv.verifyConfig(config);
```

Comment:
this could be super.verifyConfig(config) I think, if non static. and would survive renamed/different hierarchy refactors more easily.

Reply:
We could use a pattern like:

```ts
export class A extends Layer {
  constructor(config) {
    super(config);
    this.verifyConfig(config);
  }
  protected verifyConfig(config: Config) { return config.a > 0 }
}
export class B extends A {
  constructor(config) {
    super(config);
    this.verifyConfig(config);
  }
  protected verifyConfig(config: Config) {
    super.verifyConfig(config);
    return config.b > 0;
  }
}
```

However, this pattern would result in the `verifyConfig`s each being called twice.  Once because of the call path

```
B.constructor 
  super(config)
    this.verifyConfig(config)   // calls B.verifyConfig since `this` is a B
      super.verifyConfig(config)   // calls A.verifyConfig since `super` here is an A
```
and then again because of the call path
```
B.constructor 
  this.verifyConfig(config)  // calls B.verifyConfig since `this` is a B
      super.verifyConfig(config)   // calls A.verifyConfig since `super` here is an A
```

Is there a better pattern here?  Is it ok if the calls are made more than once?

Reply:
Another slightly different option is to call `super.verifyConfig` directly from the constructor.  I believe this still has problem of multiple calls.

```ts
export class A extends Layer {
  constructor(config) {
    super(config);
    super.verifyConfig(config);
    this.verifyConfig(config);
  }
  protected verifyConfig(config: Config) { return config.a > 0 }
}
export class B extends A {
  constructor(config) {
    super(config);
    super.verifyConfig(config)
    this.verifyConfig(config);
  }
  protected verifyConfig(config: Config) { return config.b > 0; }
}
```

Reply:
one other thought I had,  if keeping them static which it sounds like we have to do.  Would we prefer to add the veritfy calls to the factory methods and not the constructors?  (So we can fail early -- however it does decouple the invariant....

Reply:
Some of these abstract classes don't have factory methods.   Also, putting the call into the constructor gives us the hierarchical calls to the super.verifyConfig for free.  If we were to move the call into the factory we would have something like
```js
static leafClass(config): Layer {
    GrandParentClass.verifyConfig(config);
    ParentClass.verifyConfig(confg);
    LeafClass.verifyConfig(config);
    return new LeafClass(config);
}
```
no?